torch/distributed/_shard/checkpoint/state_dict_loader.py

```diff
@@ -132,25 +139,56 @@ def load_state_dict(
         >>> # ensure correct behavior.
         >>> my_model.load_state_dict(model_state_dict)
     """
-
-    metadata = storage_reader.read_metadata()
-    bytes_read_requests, tensor_read_requests = _reshard_and_prepare_read_request(
-        state_dict=state_dict, metadata_from_storage=metadata
-    )
-    bytes_futures = storage_reader.read_bytes(bytes_read_requests)
-    tensor_futures = storage_reader.read_tensors(tensor_read_requests)
-
-    bytes_futures.wait()
-
-    # Addtional steps are required to convert the bytes to its original type
-    # Note that this is NOT inplace,
-    # it creating a new object and replace what's in the state dict
-    for req in bytes_read_requests:
-        # Ensure the BytesIO is rewound
-        req.bytes.seek(0)
-        state_dict[req.fqn] = torch.load(req.bytes)
-
-    tensor_futures.wait()
+    is_coordinator = not dist.is_initialized() or dist.get_rank(process_group) == coordinator_rank
+
+    try:
+        metadata = storage_reader.read_metadata()
+        bytes_read_requests, tensor_read_requests = _reshard_and_prepare_read_request(
+            state_dict=state_dict, metadata_from_storage=metadata
+        )
+        bytes_futures = storage_reader.read_bytes(bytes_read_requests)
+        tensor_futures = storage_reader.read_tensors(tensor_read_requests)
+
+        bytes_futures.wait()
+
+        # Addtional steps are required to convert the bytes to its original type
+        # Note that this is NOT inplace,
+        # it creating a new object and replace what's in the state dict
+        for req in bytes_read_requests:
+            # Ensure the BytesIO is rewound
+            req.bytes.seek(0)
+            state_dict[req.fqn] = torch.load(req.bytes)
+
+        tensor_futures.wait()
+        result = None
+    except BaseException as e:
+        result = e
+
+    global_result: List[Optional[CheckpointException]] = [None]
+    if dist.is_initialized():
+        # FIXME we could do an all_to_all_object if once existed
+        all_errors = [None] * dist.get_world_size(process_group)
+
+        dist.gather_object(
```

Comment:
Don't we need to set the appropriate cuda device before the gather/broadcast object calls? If the user is expected to do this, we should document that clearly in the docs for `load_state_dict`

Reply:
I updated the doc using the text in ShardedTensor as a base - I fixed a couple of typos there as well.